File: 99.Code/PYServer/stock/high_low.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from datetime import datetime
import dal


def syn(code = ""):
    if code:
        getStockDays(code)
    else:
        sql = "SELECT code FROM Stock ORDER BY updateDate"
        for stock in dal.queryAll(sql):
            getStockDays(stock[0])


def getStockDays(code):
    sql = "SELECT * FROM StockKD WHERE code='%s' ORDER BY day" % code
    days = dal.queryAll(sql)

    today = days[-1][2]

    oldHighDays = 1
    oldHigh = float(days[-2][4])
    for i in range(1, len(days)):
        day = days[-1 - i]
        if oldHigh < float(day[4]):
            oldHighDays = i - 1
            break
        else:
            oldHighDays = i - 1
            continue

    highDays = 1
    todayHigh = float(days[-1][4])
    for i in range(0, len(days)):
        day = days[-1 - i]
        if todayHigh < float(day[4]):
            highDays = i
            break
        else:
            highDays = i
            continue

    oldLowDays = 1
    oldLow = float(days[-2][5])
    for i in range(1, len(days)):
        day = days[-1 - i]
        if float(day[5]) < oldLow:
            oldLowDays = i - 1
            break
        else:
            oldLowDays = i - 1
            continue

    lowDays = 1
    todayLow = float(days[-1][5])
    for i in range(0, len(days)):
        day = days[-1 - i]
        if float(day[5]) < todayLow:
            lowDays = i
            break
        else:
            lowDays = i
            continue

    sql = "UPDATE Stock SET oldHighDays=%d,newHighDays=%d,oldLowDays=%d,newLowDays=%d,updateDate='%s' WHERE code='%s'" % (oldHighDays, highDays, oldLowDays, lowDays, today, code)
    rowc = dal.run(sql)
    print("计算高低值", code, "--", oldHighDays, "--", highDays, "--", oldLowDays, "--", lowDays, "--", rowc)

File: 99.Code/Server/biz/stock/high_low.js
var dal = require('../../dal/main');
var http = require("http");
var iconv = require('iconv-lite');

module.exports = {
    batchStockDays: () => {
        var db = new dal();
        var sql = "SELECT * FROM StockK_Day order by code,day";
        db.query(sql)
            .then((rows) => {
                var lastCode = rows[0].code;
                var lastDays = [ rows[0] ];
                for (var i = 1; i < rows.length; i++) {
                    if (rows[i].code != lastCode){
                        StockDays.push({ code: lastCode, days: lastDays});
                        lastCode = rows[i].code;
                        lastDays = [ rows[i] ];
                    } else {
                        lastDays.push(rows[i])
                    }
                }
                StockDays.push({ code: lastCode, days: lastDays});
                calHighLow();
            })
            .catch((err) => { })
            .finally(() => { db.close(); });
    },
};

var StockDays = [];
var StockDaysIndex = 0;
var calHighLow = () => {
    var code = StockDays[StockDaysIndex].code;
    var days = StockDays[StockDaysIndex].days;

    var today = days[days.length - 1].day;

    var oldHighDays = 1;
    var oldHigh = parseFloat(days[days.length - 2].high);
    for (var i = 1; i < days.length; i++) {
        var day = days[days.length - 1 - i];
        if(oldHigh < parseFloat(day.high)){
            oldHighDays = i - 1;
            break;
        } else {
            continue;
        }
    }

    var highDays = 1;
    var todayHigh = parseFloat(days[days.length - 1].high);
    for (var i = 0; i < days.length; i++) {
        var day = days[days.length - 1 - i];
        if(todayHigh < parseFloat(day.high)){
            highDays = i;
            break;
        } else {
            continue;
        }
    }

    var oldLowDays = 1;
    var oldLow = parseFloat(days[days.length - 2].low);
    for (var i = 1; i < days.length; i++) {
        var day = days[days.length - 1 - i];
        if(parseFloat(day.low) < oldLow){
            oldLowDays = i - 1;
            break;
        } else {
            continue;
        }
    }

    var lowDays = 1;
    var todayLow = parseFloat(days[days.length - 1].low);
    for (var i = 0; i < days.length; i++) {
        var day = days[days.length - 1 - i];
        if(parseFloat(day.low) < todayLow){
            lowDays = i;
            break;
        } else {
            continue;
        }
    }
    console.log(code + "----" + oldHighDays + "----" + highDays + "----" + oldLowDays + "----" + lowDays)

    var db = new dal();
    var sql = "UPDATE Stock SET oldHighDays="+oldHighDays+",newHighDays="+highDays+",oldLowDays="+oldLowDays+",newLowDays="+lowDays+",updateDate='"+today+"' WHERE code='"+code+"'";
    db.run(sql).then(() => {
        if (StockDaysIndex >= StockDays.length - 1){
            console.log("高低值计算完成");
        } else {
            StockDaysIndex++;
            calHighLow();
        }
    }).catch((err) => { }).finally(() => { db.close(); });
};